### PR TITLE
[MIG] partner_changeset: Migrated to 10.0

### PR DESCRIPTION
**Pull request by yvaucher**
History taken from 9.0 branch

for migration change diff see commit https://github.com/OCA/partner-contact/pull/380/commits/81398af2c5b97fc426f1c1a409ceb89d3daae6ca

### CODE REVIEW

**Comment by yvaucher** [comment]
travis error doesn't seems to be related, this might be an upstream bug (Access error when installing base xml data)

restarted the job

File: partner_changeset/README.rst [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  .. image:: partner_changeset/static/src/img/rules.png
+
+* Changeset waiting for validation
+
+  .. image:: partner_changeset/static/src/img/changeset.png
+
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/134/9.0
```
Comment (cubells) [inline partner_changeset/README.rst]: It must be 10.0

File: partner_changeset/models/changeset_field_rule.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015-2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api
```
Comment (cubells) [inline partner_changeset/models/changeset_field_rule.py]: Reorder imports please: from odoo import api, fields, models

File: partner_changeset/models/res_partner.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015-2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api
```
Comment (cubells) [inline partner_changeset/models/res_partner.py]: SAme as above

File: partner_changeset/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+class ResPartner(models.Model):
+    _inherit = 'res.partner'
+
+    changeset_ids = fields.One2many(comodel_name='res.partner.changeset',
+                                    inverse_name='partner_id',
+                                    string='Changesets',
+                                    readonly=True)
+    count_pending_changesets = fields.Integer(
+        string='Changes',
+        compute='_count_pending_changesets',
```
Comment (cubells) [inline partner_changeset/models/res_partner.py:17]: Must start with _compute

File: partner_changeset/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        search='_search_count_pending_changesets')
+
+    @api.one
```
Comment (cubells) [inline partner_changeset/models/res_partner.py]: use of api.one is not recommended

File: partner_changeset/models/res_partner_changeset.py [hunk 6]
```diff
@@ -0,0 +1,515 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015-2017 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from itertools import groupby
+from lxml import etree
+from operator import attrgetter
+
+from odoo import models, fields, api, exceptions, _
```
Comment (cubells) [inline partner_changeset/models/res_partner_changeset.py]: same as above

File: partner_changeset/models/res_partner_changeset.py [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+        selection='_reference_models',
+        readonly=True,
+    )
+
+    @api.model
+    def _reference_models(self):
+        models = self.env['ir.model'].search([])
+        return [(model.model, model.name) for model in models]
+
+    @api.one
```
Comment (cubells) [inline partner_changeset/models/res_partner_changeset.py]: use of api.one is not recommended

File: partner_changeset/models/res_partner_changeset.py [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+                            for suffix in _suffix_to_types]
+    _old_value_fields = ['old_value_%s' % suffix
+                         for suffix in _suffix_to_types]
+    _new_value_fields = ['new_value_%s' % suffix
+                         for suffix in _suffix_to_types]
+    _value_fields = (_origin_value_fields +
+                     _old_value_fields +
+                     _new_value_fields)
+
+    @api.one
```
Comment (cubells) [inline partner_changeset/models/res_partner_changeset.py]: use of api.one is not recommended

File: partner_changeset/models/res_partner_changeset.py [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+    def _compute_origin_values(self):
+        field_name = self.get_field_for_type(self.field_id, 'origin')
+        if self.state == 'draft':
+            value = self.changeset_id.partner_id[self.field_id.name]
+        else:
+            old_field = self.get_field_for_type(self.field_id, 'old')
+            value = self[old_field]
+        setattr(self, field_name, value)
+
+    @api.one
```
Comment (cubells) [inline partner_changeset/models/res_partner_changeset.py]: use of api.one is not recommended

**Review by cubells — CHANGES_REQUESTED** [review]
Fast code review

**Comment by cubells** [comment]
Travis error is not related to this module.

**Review by antoniocanovas — CHANGES_REQUESTED** [review]
Work as described.
It would be good to have any signal for the end user with the pending approvals (like admin user).
If it is not implemented the end user will do the changes several times and finally will create an issue.

**Review by leemannd — APPROVED** [review]
Thanks. Just squash transifex commits

**Comment by damdam-s** [comment]
@yvaucher @leemannd @cubells can you please check it to have this one merged very soon please ?

**Comment by damdam-s** [comment]
tests fails because of this module `partner_create_by_vat`

File: partner_changeset/models/res_partner_changeset.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+
+    @api.multi
+    def apply(self):
+        self.mapped('change_ids').apply()
+
+    @api.multi
+    def cancel(self):
+        self.mapped('change_ids').cancel()
+
+    @api.multi
```
Comment (rousseldenis) [inline partner_changeset/models/res_partner_changeset.py]: This must be 'model'

File: partner_changeset/models/res_partner_changeset.py [hunk 11]
```diff
@@ -0,0 +1,516 @@
...
+            )
+            if previous_changesets:
+                raise exceptions.Warning(
+                    _('This change cannot be applied because a previous '
+                      'changeset for the same partner is pending.\n'
+                      'Apply all the anterior changesets before applying '
+                      'this one.')
+                )
+
+            partner.with_context(__no_changeset=True).write(values)
```
Comment (rousseldenis) [inline partner_changeset/models/res_partner_changeset.py]: I'm not confident in this as you create new cache at each loop iteration.

Maybe a better approach should be a other 'write' method on partner

File: partner_changeset/models/res_partner_changeset.py [hunk 12]
```diff
@@ -0,0 +1,516 @@
...
+            value = field_def.convert_to_write(record[field_name], record)
+        if field_def.type == 'many2one':
+            # store as 'reference'
+            comodel = field_def.comodel_name
+            return "%s,%s" % (comodel, value) if value else False
+        else:
+            return value
+
+    @api.multi
+    def _prepare_changeset_change(self, record, rule, field_name, value):
```
Comment (rousseldenis) [inline partner_changeset/models/res_partner_changeset.py]: This must be model

File: partner_changeset/models/res_partner_changeset.py [hunk 13]
```diff
@@ -0,0 +1,516 @@
...
+            old_field_name = self.get_field_for_type(rule.field_id, 'old')
+            # get values ready to write as expected by the changeset
+            # (for instance, a many2one is written in a reference
+            # field)
+            origin_value = self._value_for_changeset(record, field_name)
+            change[old_field_name] = origin_value
+
+        return change, pop_value
+
+    def fields_view_get(self, *args, **kwargs):
```
Comment (rousseldenis) [inline partner_changeset/models/res_partner_changeset.py]: api.model

**Review by rousseldenis — CHANGES_REQUESTED** [review]
LGTM. Code Review.

Some comments

**Comment by damdam-s** [comment]
@pedrobaeza @cubells @rousseldenis fixes done. can we go forward on this ?

**Review by cubells — APPROVED** [review]
:+1: 

LGTM

Code review and test.

Good job @damdam-s

**Review by rousseldenis — APPROVED** [review]
Code review :+1: 

@damdam-s Can you see why travis tests fail?

**Comment by cubells** [comment]
@rousseldenis @damdam-s 
Travis fails due to `partner_create_by_vat` tests.

As `partner_create_by_vat` is executed after `partner_changeset`, and `partner_changeset` tests change permission of street field, when `partner_create_by_vat` test try to change street value, this value is not changed.

So @damdam-s you must restore partner fields permissions in tests in order to avoid similar fails in future.

Do you understand?

**Comment by damdam-s** [comment]
@cubells tests corrected. 
Is this PR ok for you ?
Thanks for your time

**Comment by cubells** [comment]
@damdam-s 

I'll retest later.

**Comment by damdam-s** [comment]
@tafaRU @eLBati @pedrobaeza as PSC members, can you review this PR please :smile: ? many thanks

**Review by leemannd — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Can you squash last 3 commits before merging?

**Comment by damdam-s** [comment]
@pedrobaeza of course. done
Thanks for taking care :smiley:

**Pull request merged**